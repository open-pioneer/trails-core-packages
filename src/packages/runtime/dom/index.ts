// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export type RootNode = Document | ShadowRoot;

/**
 * Class of the <div> that contains the entire application.
 */
export const APP_ROOT_CLASS = "pioneer-root";

export function isShadowRoot(node: RootNode): node is ShadowRoot {
    return node.nodeType === Node.DOCUMENT_FRAGMENT_NODE && "host" in node;
}

export function isDocument(node: RootNode): node is Document {
    return !isShadowRoot(node);
}

/**
 * Returns the parent node for <style> elements.
 */
export function getStylesRoot(node: RootNode): ShadowRoot | HTMLHeadElement {
    return isShadowRoot(node) ? node : node.head;
}

/**
 * Creates the node that contains all other application content.
 */
export function createAppRoot(locale: string): HTMLDivElement {
    // Setup application root node in the shadow dom
    const container = document.createElement("div");
    container.classList.add("pioneer-root");
    container.style.minHeight = "100%";
    container.style.height = "100%";
    if (locale) {
        container.lang = locale;
    }
    return container;
}

/**
 * Returns the styles for host element (the web component node).
 */
export function getBuiltinStyles(rootNode: RootNode, hostElement: HTMLElement | undefined) {
    // Prevent inheritance of certain css values and normalize to display: block by default.
    // See https://open-wc.org/guides/knowledge/styling/styles-piercing-shadow-dom/
    // NOTE: layer base comes from chakra (used for css resets etc).
    const selector =
        isShadowRoot(rootNode) || !hostElement ? ":host" : hostElement.tagName.toLowerCase();
    return `@layer base { ${selector} { all: initial; display: block; } }`;
}
