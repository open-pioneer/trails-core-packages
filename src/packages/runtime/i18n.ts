// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createIntl, createIntlCache, IntlFormatters, IntlShape } from "@formatjs/intl";
import { ErrorId } from "./errors";
import { createLogger, Error } from "@open-pioneer/core";
import { ApplicationMetadata } from "./metadata";
const LOG = createLogger("runtime:i18n");

/**
 * Represents i18n info for the entire application.
 * Currently not exposed to user code.
 */
export interface AppI18n {
    /** Chosen locale */
    readonly locale: string;

    /** Supported locales from app metadata. */
    readonly supportedMessageLocales: string[];

    /** True if the locale can be used in this application (i.e. if there are any messages). */
    supportsLocale(locale: string): boolean;

    /** Given the package name, constructs a package i18n instance. */
    createPackageI18n(packageName: string): PackageIntl;
}

/**
 * Gives access to the package's i18n messages for the current locale.
 *
 * See also https://formatjs.io/docs/intl
 */
export type PackageIntl = Pick<IntlShape, "locale" | "timeZone"> & IntlFormatters<string>;

export function createPackageIntl(locale: string, messages: Record<string, string>) {
    const cache = createIntlCache();
    return createIntl(
        {
            locale,
            messages
        },
        cache
    );
}

/**
 * Initializes the application's locale and fetches the appropriate i18n messages.
 */
export async function initI18n(
    appMetadata: ApplicationMetadata | undefined,
    forcedLocale: string | undefined
): Promise<AppI18n> {
    const messageLocales = appMetadata?.locales ?? [];
    const userLocales = getBrowserLocales();
    if (LOG.isDebug()) {
        LOG.debug(
            `Attempting to pick locale for user (locales: ${userLocales.join(
                ", "
            )}) from app (supported locales: ${messageLocales.join(
                ", "
            )}) [forcedLocale=${forcedLocale}].`
        );
    }

    const i18nConfig = new I18nConfig(messageLocales);
    const { locale, messageLocale } = i18nConfig.pickSupportedLocale(forcedLocale, userLocales);

    if (LOG.isDebug()) {
        LOG.debug(`Using locale '${locale}' with messages from locale '${messageLocale}'.`);
    }

    let messages: Record<string, Record<string, string>>;
    if (messageLocales.includes(messageLocale)) {
        try {
            messages = (await appMetadata?.loadMessages?.(messageLocale)) ?? {};
        } catch (e) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to load messages for locale '${messageLocale}'.`,
                {
                    cause: e
                }
            );
        }
    }
    return {
        locale,
        supportedMessageLocales: messageLocales,
        supportsLocale(locale) {
            return i18nConfig.supportsLocale(locale);
        },
        createPackageI18n(packageName) {
            const packageMessage = messages?.[packageName] ?? {};
            return createPackageIntl(locale, packageMessage);
        }
    };
}

/** Creates an empty i18n instance, e.g. for tests. */
export function createEmptyI18n(locale = "en"): PackageIntl {
    return createPackageIntl(locale, {});
}

export interface LocalePickResult {
    /**
     * The actual locale (e.g. en-US) for number and date formatting etc.
     */
    locale: string;

    /**
     * The locale identifier to load messages for. E.g. "en".
     */
    messageLocale: string;
}

/**
 * Picks a locale for the app. Exported for tests.
 */
export class I18nConfig {
    private appLocales: string[];

    /**
     * @param appLocales Locales the app has i18n messages for (e.g. "en", "de")
     */
    constructor(appLocales: string[]) {
        this.appLocales = appLocales;
    }

    /**
     * @param forcedLocale Optional forced locale (must be satisfied)
     * @param userLocales Locales requested by the user's browser
     */
    pickSupportedLocale(forcedLocale: string | undefined, userLocales: string[]): LocalePickResult {
        const { appLocales } = this;
        // Attempt to satisfy forced locale
        if (forcedLocale) {
            const result = this.pickImpl([forcedLocale]);
            if (!result) {
                const localesList = appLocales.join(", ");
                throw new Error(
                    ErrorId.UNSUPPORTED_LOCALE,
                    `Locale '${forcedLocale}' cannot be forced because it is not supported by the application.` +
                        ` Supported locales are ${localesList}.`
                );
            }
            if (result.locale !== result.messageLocale) {
                LOG.warn(
                    `Non-exact match for forced locale '${forcedLocale}': using messages from '${result.messageLocale}'.`
                );
            }
            return result;
        }

        // Match preferred locale
        const supportedLocale = this.pickImpl(userLocales);
        if (supportedLocale) {
            return supportedLocale;
        }

        // Fallback: Most preferred locale (for dates etc.), but some of our messages
        return {
            locale: userLocales[0] ?? "en",
            messageLocale: appLocales[0] ?? "en"
        };
    }

    supportsLocale(locale: string): boolean {
        return this.pickImpl([locale]) != null;
    }

    private pickImpl(requestedLocales: string[]) {
        const appLocales = this.appLocales;
        for (const requestedLocale of requestedLocales) {
            // try exact match
            if (appLocales.includes(requestedLocale)) {
                return { messageLocale: requestedLocale, locale: requestedLocale };
            }

            // try plain language tag (e.g. "en") as fallback
            const plainLanguage = requestedLocale.match(/^([a-z]+)/i)?.[1];
            if (plainLanguage && appLocales.includes(plainLanguage)) {
                return { messageLocale: plainLanguage, locale: requestedLocale };
            }
        }
        return undefined;
    }
}

/**
 * Returns locales supported by the browser, in order of preference (preferred first).
 *
 * See also https://developer.mozilla.org/en-US/docs/Web/API/Navigator/languages
 */
export function getBrowserLocales(): string[] {
    if (window.navigator.languages && window.navigator.languages.length) {
        return Array.from(window.navigator.languages);
    }
    return [window.navigator.language];
}
