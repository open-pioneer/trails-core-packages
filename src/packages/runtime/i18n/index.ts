// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export { type AppIntl, initI18n } from "./AppIntl";
export { type PackageIntl, createPackageIntl, createEmptyPackageIntl } from "./PackageIntl";
export { type LocalePickResult, I18nConfig, getBrowserLocales } from "./pick";
