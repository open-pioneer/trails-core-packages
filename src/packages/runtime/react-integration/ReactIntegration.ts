// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { PackageContextMethods } from "@open-pioneer/runtime-react-support";
import { ComponentType, createElement } from "react";
import { Root, createRoot } from "react-dom/client";
import { RootUI } from "./RootUI";

export interface ReactIntegrationOptions {
    /** Div where react may render itself. */
    containerNode: HTMLDivElement;

    /** Component root node (e.g. shadow root), parent of `containerNode`. */
    rootNode: Node;

    /**
     * Enable or disable the global error handler.
     * Used for testing, should be `true` in production.
     */
    enableErrorHandler: boolean;
}

export interface UiStateLoading {
    type: "loading";
}

export interface UiStateError {
    type: "error";
    error: Error;
}

export interface UiStateReady {
    type: "ready";
    packageContext: PackageContextMethods;
    Component: ComponentType;
    componentProps: Record<string, unknown>;
}

export type UiState = UiStateLoading | UiStateError | UiStateReady;

export class ReactIntegration {
    private rootNode: Node;
    private reactRoot: Root;
    private enableErrorHandler: boolean;

    constructor(options: ReactIntegrationOptions) {
        this.rootNode = options.rootNode;
        this.reactRoot = createRoot(options.containerNode);
        this.enableErrorHandler = options.enableErrorHandler;
    }

    destroy() {
        this.reactRoot.unmount();
    }

    render(state: UiState) {
        this.reactRoot.render(
            createElement(RootUI, {
                rootNode: this.rootNode,
                enableErrorBoundary: this.enableErrorHandler,
                state: state
            })
        );
    }
}
