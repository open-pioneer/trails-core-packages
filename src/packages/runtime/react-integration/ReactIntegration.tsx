// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { ComponentType, StrictMode } from "react";
import { createRoot, Root } from "react-dom/client";
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { ServiceLayer } from "../service-layer/ServiceLayer";
import { PackageContext, PackageContextMethods } from "@open-pioneer/runtime-react-support";
import { PackageRepr } from "../service-layer/PackageRepr";
import { InterfaceSpec, renderInterfaceSpec } from "../service-layer/InterfaceSpec";
import { renderAmbiguousServiceChoices } from "../service-layer/ServiceLookup";
import { CustomChakraProvider } from "@open-pioneer/chakra-integration";
import { theme as defaultTrailsTheme } from "@open-pioneer/base-theme";

export interface ReactIntegrationOptions {
    packages: Map<string, PackageRepr>;
    serviceLayer: ServiceLayer;
    rootNode: HTMLDivElement;
    container: Node;
    theme: Record<string, unknown> | undefined;
}

export class ReactIntegration {
    private containerNode: Node;
    private theme: Record<string, unknown> | undefined;
    private root: Root;
    private packageContext: PackageContextMethods;

    static createForApp(options: ReactIntegrationOptions): ReactIntegration {
        const { serviceLayer, packages } = options;
        const packageContext = createPackageContext(serviceLayer, packages);
        return new ReactIntegration({ ...options, packageContext });
    }

    static createForErrorScreen(
        options: Omit<ReactIntegrationOptions, "serviceLayer" | "packages">
    ) {
        const throwError = () => {
            // TODO: Pick an error id
            throw new Error(ErrorId.INTERNAL, "Hook cannot be used within the error screen.");
        };
        const packageContext: PackageContextMethods = {
            getIntl: throwError,
            getProperties: throwError,
            getService: throwError,
            getServices: throwError
        };
        return new ReactIntegration({ ...options, packageContext });
    }

    private constructor(
        options: Omit<ReactIntegrationOptions, "serviceLayer" | "packages"> & {
            packageContext: PackageContextMethods;
        }
    ) {
        this.containerNode = options.container;
        this.theme = options.theme;
        this.root = createRoot(options.rootNode);
        this.packageContext = options.packageContext;
    }

    render(Component: ComponentType) {
        this.root.render(
            <StrictMode>
                <CustomChakraProvider
                    container={this.containerNode}
                    colorMode="light"
                    theme={this.theme ?? defaultTrailsTheme}
                >
                    <PackageContext.Provider value={this.packageContext}>
                        <Component />
                    </PackageContext.Provider>
                </CustomChakraProvider>
            </StrictMode>
        );
    }

    destroy() {
        this.root.unmount();
    }
}

function createPackageContext(
    serviceLayer: ServiceLayer,
    packages: Map<string, PackageRepr>
): PackageContextMethods {
    const getPackage = (packageName: string): PackageRepr => {
        const pkg = packages.get(packageName); // todo own error message if this.packages is undefined?
        if (!pkg) {
            throw new Error(
                ErrorId.INTERNAL,
                `Package '${packageName}' was not found in application.`
            );
        }
        return pkg;
    };

    const packageContext: PackageContextMethods = {
        getService: (packageName, interfaceName, options) => {
            const spec: InterfaceSpec = { interfaceName, ...options };
            const result = serviceLayer.getService(packageName, spec);
            if (result.type === "found") {
                return result.value.getInstanceOrThrow();
            }

            const renderedSpec = renderInterfaceSpec(spec);
            switch (result.type) {
                case "unimplemented":
                    throw new Error(
                        ErrorId.INTERFACE_NOT_FOUND,
                        `The UI of package '${packageName}' requested the unimplemented interface ${renderedSpec}.`
                    );
                case "undeclared":
                    throw new Error(
                        ErrorId.UNDECLARED_DEPENDENCY,
                        `Package '${packageName}' did not declare an UI dependency on interface ${renderedSpec}.` +
                            ` Add the dependency to the package configuration or remove the usage.`
                    );
                case "ambiguous": {
                    const renderedChoices = renderAmbiguousServiceChoices(result.choices);
                    throw new Error(
                        ErrorId.AMBIGUOUS_DEPENDENCY,
                        `The UI of package '${packageName}' requires the ambiguous interface ${renderedSpec}.` +
                            ` Possible choices are: ${renderedChoices}.`
                    );
                }
                case "unknown-package":
                    throw new Error(
                        ErrorId.MISSING_PACKAGE,
                        missingPackageMessage(packageName, interfaceName)
                    );
            }
        },
        getServices: (packageName, interfaceName) => {
            const result = serviceLayer.getServices(packageName, interfaceName);
            if (result.type === "found") {
                return result.value.map((serviceRepr) => serviceRepr.getInstanceOrThrow());
            }

            switch (result.type) {
                case "undeclared":
                    throw new Error(
                        ErrorId.UNDECLARED_DEPENDENCY,
                        `Package '${packageName}' did not declare an UI dependency on all services implementing interface '${interfaceName}'.` +
                            ` Add the dependency ("all": true) to the package configuration or remove the usage.`
                    );
                case "unknown-package":
                    throw new Error(
                        ErrorId.MISSING_PACKAGE,
                        missingPackageMessage(packageName, interfaceName)
                    );
            }
        },
        getProperties: (packageName) => {
            return getPackage(packageName).properties;
        },
        getIntl: (packageName) => {
            return getPackage(packageName).intl;
        }
    };
    return packageContext;
}

function missingPackageMessage(packageName: string, interfaceName: string) {
    return (
        `Package '${packageName}' was not found in the application's metadata while it attempted to reference the interface '${interfaceName}'. ` +
        `Check that the dependency is declared correctly in the packages that use '${packageName}'.`
    );
}
