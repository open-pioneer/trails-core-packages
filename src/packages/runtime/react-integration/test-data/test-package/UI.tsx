import { useProperties, useService, useServices } from "open-pioneer:react-hooks";

export function UIWithProperties() {
    const properties = useProperties();
    return <div className="ui">Test-UI: {properties.greeting as string}</div>
}

export function UIWithService() {
    const service = useService("test.Provider") as any;
    return <div className="ui">Test-UI: {service.value}</div>;
}

export function UIWithServices() {
    const services = useServices("test.Provider") as any[];
    return <div className="ui">Test-UI: {services.map(service => service.value).join(",")}</div>;
}
