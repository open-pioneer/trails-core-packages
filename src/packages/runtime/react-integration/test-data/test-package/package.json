{
    "name": "@open-pioneer/runtime__react_test_package"
}
