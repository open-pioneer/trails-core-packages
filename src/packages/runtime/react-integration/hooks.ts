// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Error } from "@open-pioneer/core";
import { useContext, useMemo } from "react";
import { InterfaceNameForServiceType } from "../DeclaredService";
import { ErrorId } from "../errors";
import { PackageIntl } from "../i18n";
import { PackageContext, PackageContextMethods } from "./PackageContext";

/*

    IMPORTANT
    =========

    The functions in this file are part of the public API, although they should not be called directly.
    They are imported by code generated during separate package compilation generation via:

        import { hookName } from '@open-pioneer/runtime/react-integration'

    They are also imported by the vite plugin during development and when building an application.

    The matching exports are in the neighboring `index.ts`.
    The interfaces of these hooks must remain stable so that already published packages can continue to work in the future.

    See also the implementation of @open-pioneer/build-package.

*/

export interface UseServiceOptions {
    /** An additional qualifier to disambiguate service instances that implement the same interface. */
    qualifier?: string;
}

/**
 * Returns a service that implements the given interface.
 * Receives the package name of the importing package as a string.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useServiceInternal<ServiceType = unknown>(
    packageName: string,
    interfaceName: InterfaceNameForServiceType<ServiceType>,
    options?: UseServiceOptions
): ServiceType;
export function useServiceInternal(
    packageName: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interfaceName: any,
    options?: UseServiceOptions
): unknown {
    const context = useContext(PackageContext);
    const service = useMemo(
        () =>
            checkContext("useService", context).getService(
                packageName,
                interfaceName,
                options ?? {}
            ),
        [context, packageName, interfaceName, options]
    );
    return service;
}

/**
 * Returns all services that implement the given interface.
 * Receives the package name of the importing package as a string.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useServicesInternal<ServiceType = unknown>(
    packageName: string,
    interfaceName: InterfaceNameForServiceType<ServiceType>,
    options?: UseServiceOptions
): ServiceType[];
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useServicesInternal(packageName: string, interfaceName: any): unknown[] {
    const context = useContext(PackageContext);
    const services = useMemo(
        () => checkContext("useServices", context).getServices(packageName, interfaceName),
        [context, packageName, interfaceName]
    );
    return services;
}

/**
 * Returns the properties for the given package.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function usePropertiesInternal(packageName: string): Readonly<Record<string, unknown>> {
    const context = useContext(PackageContext);
    return checkContext("useProperties", context).getProperties(packageName);
}

/**
 * Returns the i18n object for the given package.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useIntlInternal(packageName: string): PackageIntl {
    const context = useContext(PackageContext);
    return checkContext("useIntl", context).getIntl(packageName);
}

function checkContext(
    hookName: string,
    contextData: PackageContextMethods | null
): PackageContextMethods {
    if (!contextData) {
        throw new Error(
            ErrorId.INTERNAL,
            `"Failed to access package context from '${hookName}': react integration was not set up properly.`
        );
    }
    return contextData;
}
