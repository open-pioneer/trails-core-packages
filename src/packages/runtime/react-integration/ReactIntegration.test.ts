// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
/**
 * @vitest-environment happy-dom
 */
import { findByTestId, findByText } from "@testing-library/dom";
import { act } from "@testing-library/react";
import { createElement } from "react";
import { beforeEach, expect, it, MockInstance, afterEach, vi, describe } from "vitest";
import { Service, ServiceConstructor } from "../Service";
import { usePropertiesInternal, useServiceInternal, useServicesInternal } from "./hooks";
import { useTheme } from "@open-pioneer/chakra-integration";
import { PackageIntl, createEmptyI18n } from "../i18n";
import { InterfaceSpec, ReferenceSpec } from "../service-layer/InterfaceSpec";
import { PackageRepr } from "../service-layer/PackageRepr";
import { ServiceLayer } from "../service-layer/ServiceLayer";
import { ServiceRepr, createConstructorFactory } from "../service-layer/ServiceRepr";
import { ReactIntegration } from "./ReactIntegration";

// eslint-disable-next-line import/no-relative-packages
import { UIWithProperties, UIWithService, UIWithServices } from "./test-data/test-package/UI";

interface TestProvider {
    value: string;
}

beforeEach(() => {
    document.body.innerHTML = "";
});

let errorSpy!: MockInstance;
beforeEach(() => {
    errorSpy = vi.spyOn(console, "error");
});

afterEach(() => {
    vi.restoreAllMocks();
});

it("should allow access to service via react hook", async () => {
    function TestComponent() {
        const service = useServiceInternal<unknown>("test", "test.Provider") as TestProvider;
        return createElement("span", undefined, `Hello ${service.value}`);
    }

    const { wrapper, integration } = createIntegration({
        packageUiReferences: [{ interfaceName: "test.Provider" }],
        services: [
            {
                name: "Provider",
                interfaces: [{ interfaceName: "test.Provider" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST";
                }
            }
        ]
    });

    act(() => {
        integration.render(createElement(TestComponent));
    });

    const node = await findByText(wrapper, "Hello TEST");
    expect(node).toMatchSnapshot();
});

it("should get error when using undefined service", async () => {
    function TestComponent() {
        const service = useServiceInternal<unknown>(
            "test",
            "test.InterfaceThatDoesNotExist"
        ) as TestProvider;
        return createElement("span", undefined, `Hello ${service.value}`);
    }

    const { integration } = createIntegration();

    expect(() => {
        act(() => {
            integration.render(createElement(TestComponent));
        });
    }).toThrowErrorMatchingInlineSnapshot(
        `[Error: runtime:undeclared-dependency: Package 'test' did not declare an UI dependency on interface 'test.InterfaceThatDoesNotExist'. Add the dependency to the package configuration or remove the usage.]`
    );
});

it("reports a helpful error when package metadata is missing", async () => {
    // This problem can happen when a package is not correctly declared as a dependency in package.json
    // (so its metadata are not discovered by the vite plugin) but it is still imported from the application
    // where it attempts to use a service.
    function TestComponent() {
        const service = useServiceInternal<unknown>("packageMissingFromMetadata", "test.Provider");
        return String(service);
    }

    const { integration } = createIntegration();
    expect(() => {
        act(() => {
            integration.render(createElement(TestComponent));
        });
    }).toThrowErrorMatchingInlineSnapshot(
        `[Error: runtime:missing-package: Package 'packageMissingFromMetadata' was not found in the application's metadata while it attempted to reference the interface 'test.Provider'. Check that the dependency is declared correctly in the packages that use 'packageMissingFromMetadata'.]`
    );
});

it("should allow access to service with qualifier via react hook", async () => {
    function TestComponent() {
        const service = useServiceInternal<unknown>("test", "test.Provider", {
            qualifier: "foo"
        }) as TestProvider;
        return createElement("span", undefined, `Hello ${service.value}`);
    }

    const { wrapper, integration } = createIntegration({
        services: [
            {
                name: "Provider",
                interfaces: [{ interfaceName: "test.Provider", qualifier: "foo" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST";
                }
            }
        ],
        packageUiReferences: [{ interfaceName: "test.Provider", qualifier: "foo" }]
    });

    act(() => {
        integration.render(createElement(TestComponent));
    });

    const node = await findByText(wrapper, "Hello TEST");
    expect(node).toMatchSnapshot();
});

it("should deny access to service when the qualifier does not match", async () => {
    function TestComponent() {
        const service = useServiceInternal<unknown>("test", "test.Provider", {
            qualifier: "bar"
        }) as TestProvider;
        return createElement("span", undefined, `Hello ${service.value}`);
    }

    const { integration } = createIntegration({
        services: [
            {
                name: "Provider",
                interfaces: [{ interfaceName: "test.Provider", qualifier: "foo" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST";
                }
            }
        ],
        packageUiReferences: [{ interfaceName: "test.Provider", qualifier: "foo" }]
    });

    expect(() => {
        act(() => {
            integration.render(createElement(TestComponent));
        });
    }).toThrowErrorMatchingSnapshot();
});

it("should allow access to all services via react hook", async () => {
    function TestComponent() {
        const services = useServicesInternal<unknown>("test", "test.Provider") as TestProvider[];
        return createElement(
            "span",
            undefined,
            `Joined Values: ${services.map((s) => s.value).join()}`
        );
    }

    const { wrapper, integration } = createIntegration({
        services: [
            {
                name: "Provider1",
                interfaces: [{ interfaceName: "test.Provider", qualifier: "foo" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST1";
                }
            },
            {
                name: "Provider2",
                interfaces: [{ interfaceName: "test.Provider", qualifier: "bar" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST2";
                }
            },
            {
                name: "Provider3",
                interfaces: [{ interfaceName: "test.Provider", qualifier: "baz" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST3";
                }
            }
        ],
        packageUiReferences: [{ interfaceName: "test.Provider", all: true }]
    });

    act(() => {
        integration.render(createElement(TestComponent));
    });

    const node = await findByText(wrapper, /^Joined Values:/);
    expect(node).toMatchSnapshot();
});

it("should deny access to all services if declaration is missing", async () => {
    function TestComponent() {
        const services = useServicesInternal<unknown>("test", "test.Provider") as TestProvider[];
        return createElement(
            "span",
            undefined,
            `Joined Values: ${services.map((s) => s.value).join()}`
        );
    }

    const { integration } = createIntegration({
        services: []
    });

    expect(() => {
        act(() => {
            integration.render(createElement(TestComponent));
        });
    }).toThrowErrorMatchingInlineSnapshot(
        `[Error: runtime:undeclared-dependency: Package 'test' did not declare an UI dependency on all services implementing interface 'test.Provider'. Add the dependency ("all": true) to the package configuration or remove the usage.]`
    );
});

it("should be able to read properties from react component", async () => {
    function TestComponent() {
        const properties = usePropertiesInternal("test");
        return createElement("span", undefined, `Hello ${properties.name}`);
    }

    const { wrapper, integration } = createIntegration({
        packageProperties: {
            name: "USER"
        }
    });

    act(() => {
        integration.render(createElement(TestComponent));
    });

    const node = await findByText(wrapper, "Hello USER");
    expect(node).toMatchSnapshot();
});

it("should provide the autogenerated useService hook", async () => {
    const testPackageName = "@open-pioneer/runtime__react_test_package";
    const { wrapper, integration } = createIntegration({
        packageName: testPackageName,
        packageUiReferences: [{ interfaceName: "test.Provider" }],
        services: [
            {
                name: "Provider",
                interfaces: [{ interfaceName: "test.Provider" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST";
                }
            }
        ]
    });

    act(() => {
        integration.render(createElement(UIWithService));
    });

    const node = await findByText(wrapper, /^Test-UI:/);
    expect(node).toMatchSnapshot();
});

it("should provide the autogenerated useServices hook", async () => {
    const testPackageName = "@open-pioneer/runtime__react_test_package";
    const { wrapper, integration } = createIntegration({
        packageName: testPackageName,
        packageUiReferences: [{ interfaceName: "test.Provider", all: true }],
        services: [
            {
                name: "Provider1",
                interfaces: [{ interfaceName: "test.Provider" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST1";
                }
            },
            {
                name: "Provider2",
                interfaces: [{ interfaceName: "test.Provider" }],
                clazz: class Provider implements Service<TestProvider> {
                    value = "TEST2";
                }
            }
        ]
    });

    act(() => {
        integration.render(createElement(UIWithServices));
    });

    const node = await findByText(wrapper, /^Test-UI:/);
    expect(node).toMatchSnapshot();
});

it("should provide the autogenerated useProperties hook", async () => {
    const testPackageName = "@open-pioneer/runtime__react_test_package";
    const { wrapper, integration } = createIntegration({
        packageName: testPackageName,
        packageProperties: {
            greeting: "Hello World!"
        }
    });

    act(() => {
        integration.render(createElement(UIWithProperties));
    });

    const node = await findByText(wrapper, /^Test-UI:/);
    expect(node).toMatchSnapshot();
});

it("should throw error when requesting properties from an unknown package", async () => {
    const { integration } = createIntegration({
        disablePackage: true
    });

    function TestComponent() {
        const properties = usePropertiesInternal("test");
        return createElement("span", undefined, `Hello ${properties.name}`);
    }

    expect(() => {
        act(() => {
            integration.render(createElement(TestComponent));
        });
    }).toThrowErrorMatchingInlineSnapshot(
        `[Error: runtime:internal-error: Package 'test' was not found in application.]`
    );
});

it("should apply the configured chakra theme", async () => {
    const testTheme = {
        colors: {
            dummyColor: "#123456"
        }
    };
    const { integration, wrapper } = createIntegration({
        disablePackage: true,
        theme: testTheme
    });

    function TestComponent() {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const theme = useTheme() as any;
        return createElement(
            "div",
            {
                "data-testid": "test-div"
            },
            `Color: ${theme.colors.dummyColor}`
        );
    }

    act(() => {
        integration.render(createElement(TestComponent));
    });

    const node = await findByTestId(wrapper, "test-div");
    expect(node.textContent).toBe("Color: #123456");
});

describe("integration for error screen ", function () {
    it("should create an ReactIntegration for an error screen", async () => {
        const integration = ReactIntegration.createForErrorScreen({
            rootNode: document.createElement("div"),
            container: document.createElement("div"),
            theme: undefined
        });

        expect(integration).toBeInstanceOf(ReactIntegration);
    });

    it("should throw an error when trying to access a service on an error screen", async () => {
        const integration = ReactIntegration.createForErrorScreen({
            rootNode: document.createElement("div"),
            container: document.createElement("div"),
            theme: undefined
        });

        function TestComponent() {
            const service = useServiceInternal<unknown>("test", "test.Provider") as TestProvider;
            return createElement("span", undefined, `Hello ${service.value}`);
        }

        expect(() => {
            act(() => {
                integration.render(createElement(TestComponent));
            });
        }).toThrowErrorMatchingInlineSnapshot(
            `[Error: runtime:invalid-state: Hook cannot be used within the error screen.]`
        );
    });
});

interface ServiceSpec {
    name: string;
    interfaces: InterfaceSpec[];
    clazz: ServiceConstructor;
}

export interface TestIntegration {
    wrapper: HTMLDivElement;
    integration: ReactIntegration;
}

function createIntegration(options?: {
    disablePackage?: boolean;
    packageName?: string;
    packageProperties?: Record<string, unknown>;
    packageUiReferences?: ReferenceSpec[];
    i18n?: PackageIntl;
    services?: ServiceSpec[];
    theme?: Record<string, unknown>;
}): TestIntegration {
    const wrapper = document.createElement("div");
    const packages = new Map<string, PackageRepr>();
    const i18n = options?.i18n ?? createEmptyI18n();
    if (!options?.disablePackage) {
        const packageName = options?.packageName ?? "test";
        const services =
            options?.services?.map((spec) => {
                return new ServiceRepr({
                    name: spec.name,
                    packageName,
                    interfaces: spec.interfaces,
                    factory: createConstructorFactory(spec.clazz),
                    intl: i18n
                });
            }) ?? [];
        packages.set(
            packageName,
            new PackageRepr({
                name: packageName,
                intl: i18n,
                properties: options?.packageProperties,
                uiReferences: options?.packageUiReferences,
                services
            })
        );
    }

    const serviceLayer = new ServiceLayer(Array.from(packages.values()));
    serviceLayer.start();

    const integration = ReactIntegration.createForApp({
        container: wrapper,
        rootNode: wrapper,
        theme: options?.theme,
        packages,
        serviceLayer
    });
    return { integration, wrapper };
}
