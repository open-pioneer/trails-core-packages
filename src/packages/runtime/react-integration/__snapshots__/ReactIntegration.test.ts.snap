// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should allow access to all services via react hook 1`] = `
<span>
  Joined Values: TEST1,TEST2,TEST3
</span>
`;

exports[`should allow access to service via react hook 1`] = `
<span>
  Hello TEST
</span>
`;

exports[`should allow access to service with qualifier via react hook 1`] = `
<span>
  Hello TEST
</span>
`;

exports[`should be able to read properties from react component 1`] = `
<span>
  Hello USER
</span>
`;

exports[`should deny access to service when the qualifier does not match 1`] = `[Error: runtime:undeclared-dependency: Package 'test' did not declare an UI dependency on interface 'test.Provider' (qualifier: 'bar'). Add the dependency to the package configuration or remove the usage.]`;

exports[`should provide the autogenerated useProperties hook 1`] = `
<div
  class="ui"
>
  Test-UI: 
  Hello World!
</div>
`;

exports[`should provide the autogenerated useService hook 1`] = `
<div
  class="ui"
>
  Test-UI: 
  TEST
</div>
`;

exports[`should provide the autogenerated useServices hook 1`] = `
<div
  class="ui"
>
  Test-UI: 
  TEST1,TEST2
</div>
`;
